### PR TITLE
batches: fix commit signing with changeset forks

### PR DESCRIPTION
**Pull request by BolajiOlajide**
[Context](https://sourcegraph.slack.com/archives/C04UV9ZVBB2/p1696374182439959)

The customer reported an issue where a Sourcegraph instance configured with a GitHub app for Batch Changes couldn't sign commits on forked changesets even when the GitHub app was granted access to the forked repo. 

This happens because while creating an authenticator used to create the signed commits, we always assume the repository to find the GitHub app installation for is the original repository where the Batch CHange was executed.

https://github.com/sourcegraph/sourcegraph/blob/main/internal/batches/sources/sources.go#L161

In this PR, I changed the approach and instead checked if the changeset is pushed to a fork, then figured out the namespace where the forked changeset is created and used that to get the GitHub app installation record.

CleanShot 2023-10-11 at 11 49 50@2x

## Test plan

* MAnual testing

- Add a GitHub app for BAtch Changes commit signing. Ensure you give it access to the forked repository of your choice.
- Add `batchChanges.enforceForks` to your site config, and create a batch change. 
- The changeset should be verified and created in your repository fork once published.

* Updated unit tests

### CODE REVIEW

File: internal/batches/sources/sources.go [hunk 1]
```diff
@@ -541,7 +570,7 @@ func getCloneURL(repo *types.Repo) (*vcs.URL, error) {
 		parsedURLs = append(parsedURLs, parsedURL)
 	}
 
-	sort.SliceStable(parsedURLs, func(i, j int) bool {
+	sort.SliceStable(parsedURLs, func(i, _ int) bool {
 		return !parsedURLs[i].IsSSH()
 	})
```
Comment (camdencheek) [inline internal/batches/sources/sources.go:575]: unrelated to this change, but I don't think this sort is guaranteed to work as intended. The thunk passed to `SliceStable` is a "less" function, so, for example, if neither `parsedURLs[i]` and `parsedURLs[j]` are SSH URLs, `i` will be sorted before `j` even though they are actually "equal" by this comparison.

```suggestion
	sort.SliceStable(parsedURLs, func(i, _ int) bool {
		return !parsedURLs[i].IsSSH() && parsedURLs[j].IsSSH()
	})
```

File: internal/github_apps/store/store.go [hunk 2]
```diff
@@ -314,10 +314,10 @@ func (s *gitHubAppsStore) GetInstallID(ctx context.Context, appID int, account s
 		FROM github_app_installs
 		JOIN github_apps ON github_app_installs.app_id = github_apps.id
 		WHERE github_apps.app_id = %s
-		AND github_app_installs.account_login = %s
+		AND LOWER(github_app_installs.account_login) = %s
 		-- We get the most recent installation, in case it's recently been removed and readded and the old ones aren't cleaned up yet.
 		ORDER BY github_app_installs.id DESC LIMIT 1
-		`, appID, account)
+		`, appID, strings.ToLower(account))
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply internal/github_apps/store/store.go:320]: Not really.

The original implementation assumed the namespace will always be searched as is (in whatever case it is), however, when we create a fork, the namespace is usually lowercase so It didn't match the name stored in the database.

File: internal/batches/sources/sources.go (hunk 1 above)
Comment (BolajiOlajide) [reply internal/batches/sources/sources.go:575]: Oh, nice catch.  I didn't know that.
Thanks for flagging.

File: internal/batches/sources/sources.go [hunk 3]
```diff
@@ -159,10 +160,38 @@ func (s *sourcer) ForChangeset(ctx context.Context, tx SourcerStore, ch *btypes.
...
+					return nil, errors.Wrap(err, "getting url for forked changeset")
+				}
+
+				owner, _, err = github.SplitRepositoryNameWithOwner(strings.TrimPrefix(u.Path, "/"))
+				if err != nil {
+					return nil, errors.Wrap(err, "getting owner from repo name")
+				}
+			}
+		} else {
+			repoMetadata := repo.Metadata.(*github.Repository)
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply internal/batches/sources/sources.go:190]: Can do!

File: internal/github_apps/store/store.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (camdencheek) [reply internal/github_apps/store/store.go:320]: Got it! Thanks

**Pull request merged**